File: clojure/02-variables/variables_types.clj
;; type returns the Java class of a value
(println "Type of 42:" (type 42))
(println "Type of 3.14:" (type 3.14))
(println "Type of 22/7:" (type 22/7))
(println "Type of \"hello\":" (type "hello"))
(println "Type of true:" (type true))
(println "Type of nil:" (type nil))
(println "Type of :keyword:" (type :keyword))
(println "Type of [1 2 3]:" (type [1 2 3]))
(println "Type of {\"a\" 1}:" (type {"a" 1}))

;; Predicate functions for type checking
(println)
(println "integer? 42:" (integer? 42))
(println "float? 3.14:" (float? 3.14))
(println "ratio? 22/7:" (ratio? 22/7))
(println "string? \"hi\":" (string? "hi"))
(println "keyword? :x:" (keyword? :x))
(println "nil? nil:" (nil? nil))
(println "nil? false:" (nil? false))
(println "boolean? true:" (boolean? true))
(println "vector? [1 2]:" (vector? [1 2]))
(println "map? {:a 1}:" (map? {:a 1}))
(println "seq? '(1 2):" (seq? '(1 2)))

;; Type conversions
(println)
(println "int->float:" (float 42))
(println "float->int:" (int 3.9))       ; truncates
(println "string->int:" (Integer/parseInt "123"))
(println "int->string:" (str 456))
(println "keyword->string:" (name :hello))
(println "string->keyword:" (keyword "world"))
(println "number->ratio:" (rationalize 0.1))  ; exact rational

File: rust/02-variables/variables_types.rs
fn main() {
    // &str: a string slice (borrowed reference, stored in the binary)
    // String: heap-allocated, owned string
    let greeting: &str = "Hello, Rust!";
    let owned: String = String::from("I own this string");
    println!("{}", greeting);
    println!("{}", owned);

    // Shadowing: rebind a name with a new value — or a completely new type.
    // This is different from mutation: each 'let' creates a new binding.
    let spaces = "   ";
    let spaces = spaces.len();  // 'spaces' is now a usize, not a &str
    println!("Spaces: {}", spaces);

    // Type casting uses the 'as' keyword. Rust never coerces silently.
    let integer: i32 = 42;
    let as_float: f64 = integer as f64;
    let truncated: i32 = 3.99_f64 as i32;  // truncates toward zero, does not round
    println!("i32 as f64: {}", as_float);
    println!("3.99 as i32: {}", truncated);

    // Tuples group values of different types. Access elements with .0, .1, etc.
    let point: (f64, f64) = (1.5, 2.7);
    let rgb: (u8, u8, u8) = (255, 128, 0);
    println!("Point: ({}, {})", point.0, point.1);
    println!("RGB: ({}, {}, {})", rgb.0, rgb.1, rgb.2);

    // Arrays hold a fixed number of values of the same type.
    // The type annotation [T; N] specifies element type and length.
    let primes: [i32; 5] = [2, 3, 5, 7, 11];
    println!("First prime: {}", primes[0]);
    println!("Array length: {}", primes.len());
}
